### PR TITLE
Pdmc2

### PR DESCRIPTION
**Pull request by traiansf**
This push request is implementing some refactorings in the pushdown model checking algorithm (separated Post automaton generation from head reachability graph), as well as adding more functionality towards generating witnesses,  Currently it can compute a path to any reachable head. 

@dwightguth: please take a look at it.
@dlucanu: I will need your "conceptual" approval for merging this code.

### CODE REVIEW

**Comment by dlucanu** [comment]
It looks ok for me.

File: src/javasources/KTool/src/org/kframework/backend/pdmc/automaton/BasicAutomaton.java [hunk 1]
```diff
@@ -123,6 +127,39 @@ public String toString() {
...
+     * @param finalState  the final state to reach
+     * @return a list of transitions describing a path from initialState to finalState or null if there is no such list.
+     */
+    public Deque<Transition<State, Alphabet>> getPath(State initialState, Alphabet initialLetter, State finalState) {
+        // We seed the algorithm with the initialState and those reachable  from it following initialLetter
+        Deque<State> toProcess = new ArrayDeque<>();
+        Map<State, Transition<State, Alphabet>> considered = new HashMap<>();
+        considered.put(initialState, null);
+
+        for (Transition<State, Alphabet> transition : getTransitions(initialState, initialLetter)) {
```
Comment (dwightguth) [inline src/javasources/KTool/src/org/kframework/backend/pdmc/automaton/BasicAutomaton.java]: I'm not comfortable with this block of code being duplicated from the extremely similar one in the method below this. Can you find a way to keep this modular?

File: src/javasources/KTool/src/org/kframework/backend/pdmc/pda/ConfigurationHead.java [hunk 2]
```diff
@@ -1,5 +1,8 @@
 package org.kframework.backend.pdmc.pda;
 
+import org.kframework.backend.java.util.Utils;
```
Comment (dwightguth) [inline src/javasources/KTool/src/org/kframework/backend/pdmc/pda/ConfigurationHead.java:3]: I don't understand what the changes in this file are doing...

File: src/javasources/KTool/src/org/kframework/backend/pdmc/pda/buchi/BuchiPushdownSystem.java [hunk 3]
```diff
@@ -35,6 +35,7 @@
     private PushdownSystemInterface<Control, Alphabet> pds;
     private PromelaBuchi ba;
     private Evaluator<ConfigurationHead<Control, Alphabet>> atomEvaluator;
+    java.util.Collection<Rule<Pair<Control, BuchiState>, Alphabet>> rules= null;
```
Comment (dwightguth) [inline src/javasources/KTool/src/org/kframework/backend/pdmc/pda/buchi/BuchiPushdownSystem.java]: Is there a reason you're not importing Collection?

File: src/javasources/KTool/test/org/kframework/backend/pdmc/pda/buchi/BuchiPushdownSystemTest.java [hunk 4]
```diff
@@ -15,21 +15,22 @@
 
 
 /**
- * Created by Traian on 04.02.2014.
+ * @author TraianSF
  */
 public class BuchiPushdownSystemTest {
     @Test
-    public void testProduct() throws Exception {
+    public void testSimpleTrue() throws Exception {
```
Comment (dwightguth) [inline src/javasources/KTool/test/org/kframework/backend/pdmc/pda/buchi/BuchiPushdownSystemTest.java:22]: Why was this test rewritten?

File: src/javasources/KTool/test/org/kframework/backend/pdmc/pda/buchi/BuchiPushdownSystemTest.java [hunk 5]
```diff
@@ -235,5 +251,13 @@ public void testMarcelloFalse() throws Exception {
 
         System.err.println("\n\n\n----Strongly Connected Components----");
         System.err.println(repeatedHeads.getSCCSString());
+
+        TarjanSCC counterExampleGraph = bpsTool.getCounterExample();
+        if (counterExampleGraph == null) {
```
Comment (dwightguth) [inline src/javasources/KTool/test/org/kframework/backend/pdmc/pda/buchi/BuchiPushdownSystemTest.java]: This should always be either true or false, so please assert which one you expect.

File: src/javasources/KTool/test/org/kframework/backend/pdmc/pda/buchi/BuchiPushdownSystemTest.java [hunk 6]
```diff
@@ -102,6 +106,15 @@ public void testProduct1() throws Exception {
 
         System.err.println("\n\n\n----Strongly Connected Components----");
         System.err.println(repeatedHeads.getSCCSString());
+
```
Comment (dwightguth) [inline src/javasources/KTool/test/org/kframework/backend/pdmc/pda/buchi/BuchiPushdownSystemTest.java:109]: I need you to assert the correctness of the repeated heads graphs and counterexamples in all the tests in this class.

File: src/javasources/KTool/test/org/kframework/backend/pdmc/pda/buchi/BuchiPushdownSystemTest.java [hunk 7]
```diff
@@ -102,6 +106,15 @@ public void testProduct1() throws Exception {
...
+        TarjanSCC<ConfigurationHead<Pair<String, BuchiState>, String>, BuchiPushdownSystemTools.LabelledAlphabet<String, String>> counterExampleGraph = bpsTool.getCounterExample();
+        Assert.assertNotNull("Property is false => counterexample exists", counterExampleGraph);
+        System.err.println("\n\n\n----CounterExample Graph----");
+        System.err.println(counterExampleGraph.toString());
+        System.err.println("\n\n\n----Reachability paths for vertices in the CounterExample Graph----");
+        for (ConfigurationHead<Pair<String, BuchiState>, String> head : counterExampleGraph.getVertices()) {
+            System.err.println(bpsTool.getReachableConfiguration(head).toString());
```
Comment (dwightguth) [inline src/javasources/KTool/test/org/kframework/backend/pdmc/pda/buchi/BuchiPushdownSystemTest.java:116]: This needs to be asserted too.

**Comment by dwightguth** [comment]
I had a lot of trouble trying to follow the algorithmic changes you made. If you can provide me with reason to believe that your test coverage is sufficient to cover all the changes you made, and that your changes are purely refactorings, I can approve this as long as you make the changes I suggested. But the problem right now is that there are significant pieces missing from the test coverage so I have trouble determining whether or not this refactoring introduced any bugs.

**Comment by dlucanu** [comment]
Dear Dwight,

We, Traian and me, discussed these changes and we agreed on them (from the algorithmic point of view). The implementation of the algorithm is quite complex and it is too early to have a set tests covering all the cases. We have a Maude implementation of the algorithm  and we compare the results of the tools in order to  be sure that the main functionality is that expected. This code does not interact with the other functionalities of the tool, so there is no the risk to introduce bugs in the tool. 
We hope to reach soon that point when we can write a test suite covering the code. Now our focus is on the main functionality.

**Comment by dwightguth** [comment]
I'm not sure I understand. Either this code works and generates the correct output you are expecting from the algorithm, or it doesn't. If it does, then your tests should assert this fact. If it doesn't, why was this code written and proposed to be committed?

**Comment by dlucanu** [comment]
As I said, the algorithm is complex and its functionality is not fully implemented . We do it incrementally and I think it is a good policy to commit after each iteration.

**Comment by dwightguth** [comment]
But you're implementing some set of functionality in each commit, correct?
If so, then that functionality should be tested in that same commit.

File: src/javasources/KTool/src/org/kframework/backend/pdmc/pda/buchi/BuchiPushdownSystem.java [hunk 8]
```diff
@@ -91,31 +92,25 @@ public boolean isFinal(Pair<Control, BuchiState> state) {
     }
 
 
-
-
-    @Override
-    public String toString() {
+    public java.util.Collection<Rule<Pair<Control, BuchiState>, Alphabet>> getRules() {
```
Comment (dwightguth) [inline src/javasources/KTool/src/org/kframework/backend/pdmc/pda/buchi/BuchiPushdownSystem.java]: javadoc

File: src/javasources/KTool/src/org/kframework/backend/pdmc/automaton/BasicAutomaton.java [hunk 9]
```diff
@@ -15,6 +15,10 @@
  * @author Traian
  */
 public class BasicAutomaton<State, Alphabet> implements AutomatonInterface<State, Alphabet> {
+    public Map<TransitionIndex<State, Alphabet>, Set<Transition<State, Alphabet>>> getDeltaIndex() {
```
Comment (dwightguth) [inline src/javasources/KTool/src/org/kframework/backend/pdmc/automaton/BasicAutomaton.java:30]: javadoc

File: src/javasources/KTool/src/org/kframework/backend/pdmc/pda/buchi/BuchiPushdownSystemTools.java [hunk 10]
```diff
@@ -109,62 +114,32 @@ public BuchiPushdownSystemTools(BuchiPushdownSystemInterface<Control, Alphabet>
...
-    public PAutomaton<PAutomatonState<Pair<Control, BuchiState>, Alphabet>,
-            LabelledAlphabet<Control, Alphabet>> getPostStar() {
+    public PAutomaton<PAutomatonState<Pair<Control, BuchiState>, Alphabet>, Alphabet> getPostStar() {
         if (postStar == null)
             compute();
         return postStar;
     }
 
-    public TarjanSCC<ConfigurationHead<Pair<Control, BuchiState>, Alphabet>, Boolean> getRepeatedHeadsGraph() {
+    public TarjanSCC<ConfigurationHead<Pair<Control, BuchiState>, Alphabet>, LabelledAlphabet<Control, Alphabet>> getRepeatedHeadsGraph() {
```
Comment (dwightguth) [inline src/javasources/KTool/src/org/kframework/backend/pdmc/pda/buchi/BuchiPushdownSystemTools.java:165]: javadoc

File: src/javasources/KTool/src/org/kframework/backend/pdmc/pda/buchi/BuchiPushdownSystemTools.java [hunk 11]
```diff
@@ -109,62 +114,32 @@ public BuchiPushdownSystemTools(BuchiPushdownSystemInterface<Control, Alphabet>
...
         if (repeatedHeadsGraph == null)
             compute();
         return repeatedHeadsGraph;
     }
 
+    public TarjanSCC<ConfigurationHead<Pair<Control, BuchiState>, Alphabet>, LabelledAlphabet<Control, Alphabet>> getCounterExample() {
```
Comment (dwightguth) [inline src/javasources/KTool/src/org/kframework/backend/pdmc/pda/buchi/BuchiPushdownSystemTools.java]: javadoc

File: src/javasources/KTool/src/org/kframework/backend/pdmc/pda/buchi/BuchiPushdownSystemTools.java [hunk 12]
```diff
@@ -302,6 +263,135 @@ public boolean isEpsilon(LabelledAlphabet<Control, Alphabet> gamma) {
...
+                    }
+                }
+            }
+        }
+
+        computeCounterExample();
+
+    }
+
+    private Collection<ConfigurationHead<Pair<Control, BuchiState>, Alphabet>> getReachableHeads(PAutomaton<PAutomatonState<Pair<Control, BuchiState>, Alphabet>, Alphabet> postStar) {
```
Comment (dwightguth) [inline src/javasources/KTool/src/org/kframework/backend/pdmc/pda/buchi/BuchiPushdownSystemTools.java:342]: javadoc

File: src/javasources/KTool/src/org/kframework/backend/pdmc/pda/graph/TarjanSCC.java [hunk 13]
```diff
@@ -127,6 +136,19 @@ public String getSCCSString() {
         return result.toString();
     }
 
+    public TarjanSCC<Data, Label> getSubgraph(Collection<TarjanSCCVertex> vertices) {
```
Comment (dwightguth) [inline src/javasources/KTool/src/org/kframework/backend/pdmc/pda/graph/TarjanSCC.java:144]: javadoc

**Comment by dwightguth** [comment]
Please refer to my email titled "IMPORTANT: Conceptual Atomicity" because it affects code reviews like this one in the future. Again, I am okay with moving this forward for now, however, it will not be an accepted pattern again in the future.

Also I would like to reiterate my point from earlier: either your code generates some specific functionally correct result, or it doesn't. If it does, you need to assert the correctness of that result. If it doesn't, you need to do more work to persuade me this is ready for code review.

File: src/javasources/KTool/src/org/kframework/backend/pdmc/automaton/BasicAutomaton.java (hunk 1 above)
Comment (traiansf) [inline src/javasources/KTool/src/org/kframework/backend/pdmc/automaton/BasicAutomaton.java]: Thanks!  I had already factored part of it but was not sure how to factor this one.  Finally figured it out, prompted by your uncomfortably (which I also shared).

File: src/javasources/KTool/src/org/kframework/backend/pdmc/pda/ConfigurationHead.java (hunk 2 above)
Comment (traiansf) [inline src/javasources/KTool/src/org/kframework/backend/pdmc/pda/ConfigurationHead.java:3]: They were suggested by IntelliJ as supposedly getting rid of a couple more of unchecked warnings.  I can revert them if you don't think they are useful, though.

File: src/javasources/KTool/src/org/kframework/backend/pdmc/pda/buchi/BuchiPushdownSystem.java (hunk 3 above)
Comment (traiansf) [inline src/javasources/KTool/src/org/kframework/backend/pdmc/pda/buchi/BuchiPushdownSystem.java]: Apparently I was importing kil.Collection (for no particular reason).  I've removed that part.

File: src/javasources/KTool/test/org/kframework/backend/pdmc/pda/buchi/BuchiPushdownSystemTest.java (hunk 4 above)
Comment (traiansf) [inline src/javasources/KTool/test/org/kframework/backend/pdmc/pda/buchi/BuchiPushdownSystemTest.java:22]: I wanted to emphasize that the two tests are dual to each-other.  furthermore, when originally written it was testing only the product of the automaton.  now it also tests whether the formula holds.

File: src/javasources/KTool/test/org/kframework/backend/pdmc/pda/buchi/BuchiPushdownSystemTest.java (hunk 7 above)
Comment (traiansf) [inline src/javasources/KTool/test/org/kframework/backend/pdmc/pda/buchi/BuchiPushdownSystemTest.java:116]: Ok, but what would you like to assert here? there might be multiple counterexamples and multiple paths.

**Comment by traiansf** [comment]
@dwightguth: I think I'm convinced.  THese files should have not been into trunk to begin with as they are still very experimental.

Therefore my plan is to remove them completely to another branch, then integrate them one by one with tests and everything.  How about that?

**Comment by dwightguth** [comment]
If by "one by one" you mean a component at a time insofar as each component
is complete in itself and has some well defined
inputs/outputs/structure/etc, then yes, that makes sense. But bear in mind
it needs to be done in such a way that the amount of code I have to review
at once is manageable. And the best way to do that is to provide me with
changes that implement discrete features.

**Comment by traiansf** [comment]
Yes, I will do that.

For now I'll close this pull request and create another one removing all
pdmc-related stuff from master.

2014-03-20 18:07 GMT+02:00 dwightguth notifications@github.com:

**Pull request closed without merging**